storage/grpc_metrics.go

```diff
@@ -0,0 +1,161 @@
+			Name: string(m),
+		}, sdkmetric.Stream{Name: strings.ReplaceAll(string(m), ".", "/")}))
+	}
+	return views
+}
+
+func metricFormatter(m metricdata.Metrics) string {
+	return metric_format_prefix + m.Name
+}
+
+func gcpAttributeExpectedDefaults() []attribute.KeyValue {
+	return []attribute.KeyValue{
+		{Key: "location", Value: attribute.StringValue(mr_location_label_default)},
+		{Key: "cloud_platform", Value: attribute.StringValue(mr_cloud_platform_label_default)},
+		{Key: "host_id", Value: attribute.StringValue(mr_host_id_label_default)}}
+}
+
+func getPreparedResourceUsingGCPDetector(ctx context.Context, project string) (*resource.Resource, error) {
+	gcpDetector := []resource.Option{resource.WithDetectors(gcp.NewDetector())}
+	return getPreparedResource(ctx, project, gcpDetector)
+}
+
+// Added to help with tests
+func getPreparedResource(ctx context.Context, project string, resourceOptions []resource.Option) (*resource.Resource, error) {
+	detectedAttrs, err := resource.New(ctx, resourceOptions...)
+	if err != nil {
+		return nil, err
+	}
+	s := detectedAttrs.Set()
+	updates := []attribute.KeyValue{}
+	for _, kv := range gcpAttributeExpectedDefaults() {
+		if val, present := s.Value(kv.Key); !present || val.AsString() == "" {
+			updates = append(updates, attribute.KeyValue{Key: kv.Key, Value: kv.Value})
+		}
+	}
+	return resource.New(
+		ctx,
+		resource.WithAttributes(
+			attribute.KeyValue{Key: "gcp.resource_type", Value: attribute.StringValue(monitored_resource_name)},
+			attribute.KeyValue{Key: "instance_id", Value: attribute.StringValue(uuid.New().String())},
+			attribute.KeyValue{Key: "project_id", Value: attribute.StringValue(project)},
+			attribute.KeyValue{Key: "api", Value: attribute.StringValue("grpc")},
+		),
+		resource.WithAttributes(detectedAttrs.Attributes()...),
+		// Last duplicate key / value wins
+		resource.WithAttributes(updates...),
+	)
+}
+
+// TODO: format errors emitted here.
+func gRPCMetricProvider(ctx context.Context, project string) (*sdkmetric.MeterProvider, error) {
+	exporter, err := mexporter.New(
+		mexporter.WithProjectID(project),
+		mexporter.WithMetricDescriptorTypeFormatter(metricFormatter),
+		mexporter.WithCreateServiceTimeSeries(),
+		mexporter.WithMonitoredResourceDescription(monitored_resource_name, []string{"project_id", "location", "cloud_platform", "host_id", "instance_id", "api"}))
+	if err != nil {
+		return nil, err
+	}
+	preparedResource, err := getPreparedResourceUsingGCPDetector(ctx, project)
+	if err != nil {
+		return nil, err
+	}
+	allMetrics := append(getMetricsEnabledByDefault(), getMetricsToEnable()...)
+	metricViews := getViewMasks(allMetrics)
+	provider := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(exporter)),
+		sdkmetric.WithResource(preparedResource),
+		sdkmetric.WithView(metricViews...),
+	)
+	return provider, nil
+}
+
+func togRPCDialOption(provider *sdkmetric.MeterProvider) option.ClientOption {
```

Comment:
nit: name seems misleading, maybe `clientOptionFromMeterProvider`. Or just do this inline since the function is only called once.